[FIX] web: in search view, use field modifiers to determine if filter is invisible

xavieralt opened the pull request:
For search view with filter having complex "invisible" attribute, like:

```xml
<search>
   ...
   <filter name="myfilter" string="My Filter" invisible="context.get('hide_myfilter')"/>
   ...
</search>
```

In such case, we must rely on the "modifiers" attribute set by the ir.ui.view
post-processing as the result can depends on the context

OPW-2243639

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

xavieralt commented:
Hi @aab-odoo, may I request your review on this PR ? Thanks!

aab-odoo commented:
ping @Arcasias

xavieralt commented:
Hi @Arcasias, did you have time to review this PR ? Thanks

Arcasias commented on addons/web/static/src/js/views/control_panel/control_panel_view.js:
```diff
@@ -177,7 +177,7 @@ var ControlPanelView = Factory.extend({
                                 attrs.domain ||
                                 'Ω';
-        if (attrs.invisible) {
+        if (JSON.parse(attrs.modifiers || '{}').invisible) {
```
If I'm understanding the fix correctly: we only use the `modifiers` attribute keys and the `invisible` attribute is rendered pointless, is that right? If so shouldn't it be removed in this PR?

xavieralt replied:
@Arcasias `modifiers` are parsed server-side in ir.ui.view [`postprocess`](https://github.com/odoo/odoo/blob/13.0/odoo/addons/base/models/ir_ui_view.py#L898) and `invisible` attribute is kept in that process (in case the it need to be re-evaluated later on and is usefull for debugging [ex: from fields_view_get()] )

xavieralt replied:
@Arcasias Note: The regression windows is between 12.0 and saas-12.1

In v12.0 this is fine in [search_input.js `visible`](https://github.com/odoo/odoo/blob/12.0/addons/web/static/src/js/views/search/search_inputs.js#L83-L85), but in saas-12.1 it seems to be lost after https://github.com/odoo/odoo/commit/40dd121938552d3f88289d2e5fcd1d57a0dd01bb

xavieralt replied:
@Arcasias I've removed the "invisible" attributes on the test, let me know if there anything else to change. Thanks!

xavieralt commented:
@aab-odoo For the test I removed the "invisible" as suggested by @Arcasias, see previous version with was still having it: https://github.com/odoo/odoo/commit/d8e0550e12da586309a5650ca0129892351acc9c. I'll re-introduce and fine-tune it.

As for v12.0, I'll test again.

Note: And *No* I'm not trying to introduce new features, this was detected during migration of customer code from v10.0 to v13.0.

xavieralt commented:
@aab-odoo @Arcasias 

I've reworked the test to:

- added one explicit `invisible` attrs using `context`

- remove the `modifiers` attributes, the `invisible` are automatically transfered to `modifiers` by the mock server fields_view_get()
  Note: It case it evaluate to `false` invisible modifiers is actually [removed](https://github.com/odoo/odoo/blob/13.0/addons/web/static/tests/helpers/mock_server.js#L302) and well as the `modifiers` [attribute itself](https://github.com/odoo/odoo/blob/13.0/addons/web/static/tests/helpers/mock_server.js#L306-L308)

- switch to `viewOptions = {context: {...}}` instead of `context: {...}` as the mock fields_view_get() only consider those [viewOptions.context](https://github.com/odoo/odoo/blob/13.0/addons/web/static/tests/helpers/mock_server.js#L88) for context (and it's needed to correctly evalute the `modifiers`)

Let me know if you need more change from me.

Note: I've created a separate PR to fix the issue in v12.0: https://github.com/odoo/odoo/pull/52465

xavieralt commented:
Will be fixed in related forward-port #53397
